### PR TITLE
Fix: Realtime transport should fire events on a serial queue

### PR DESCRIPTION
**Pull request by ricardopereira**
It wasn't firing transport events when the Realtime object was created on a background queue.
UPDATE: Rest was with the same problem.

### CODE REVIEW

File: Source/ARTRealtime.m [hunk 1]
```diff
@@ -381,21 +391,29 @@ - (void)transitionSideEffects:(ARTConnectionStateChange *)stateChange {
...
+- (_Nonnull dispatch_semaphore_t)unlessStateChangesBefore:(NSTimeInterval)deadline do:(void(^)())callback __attribute__((warn_unused_result)) {
+    // Defer until next event loop execution so that any event emitted in the current one doesn't cancel the timeout.
     ARTRealtimeConnectionState state = self.connection.state;
-    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0), dispatch_get_main_queue(), ^{
+    // Timeout should be dispatched after current event.
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0), _eventQueue, ^{
+        // Wait until the current event is done.
+        dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+        ARTEventListener *listener = [_internalEventEmitter once:^(ARTConnectionStateChange *change) {
```
Comment (tcard) [inline Source/ARTRealtime.m]: Why do we always create the listener, ie. why the condition on line 409 has been moved below this?

Comment (ricardopereira) [inline Source/ARTRealtime.m]: @tcard Ah, that's wrong. Will fix it. 👍

Comment (ricardopereira) [inline Source/ARTRealtime.m]: Done 380e6c7.

**Comment by ricardopereira** [comment]
Test added a1a77d1.

**Comment by ricardopereira** [comment]
:/

```
Failing tests:
-[ARTRealtimePresenceTest testLeaveAndGet]
-[RealtimeClientChannel Channel__attach__should_transition_the_channel_state_to_FAILED_if_ATTACHED_ProtocolMessage_is_not_received()]
-[RealtimeClientConnection Connection__basic_operations_should_work_simultaneously()]
-[RealtimeClientConnection Connection__close__if_CONNECTING__do_the_operation_once_CONNECTED()]
-[RealtimeClientConnection Connection__connection_request_fails__connection_attempt_fails_for_any_recoverable_reason()]
-[RealtimeClientPresence Presence__leave__should_leave_the_current_client_from_the_channel_and_the_data_will_be_updated_with_the_value_provided()]
** TEST FAILED **
```

I'll check what's wrong.

File: Scripts/pre-commit.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+set -eu
```
Comment (mattheworiordan) [inline Scripts/pre-commit.sh]: Did you mean to commit this?

Comment (ricardopereira) [inline Scripts/pre-commit.sh]: Yes, it helps a lot when I only run one spec and I don't want to commit that to the server.

**Comment by ricardopereira** [comment]
@mattheworiordan I already added a test: a1a77d1.

**Comment by ricardopereira** [comment]
Can I squash?

**Comment by mattheworiordan** [comment]
@ricardopereira but https://github.com/ably/ably-ios/commit/a1a77d1efcbbb4d18d86948aec53f3ba192c14f4 is in the Examples folder?  What has that to do with tests?  The name of that file btw is pretty horrific `Examples/Tests/TestsTests/TestsTests.swift`!  Has that got anything to do with Tests? 😏

File: Scripts/pre-commit.sh (hunk 2 above)
Comment (mattheworiordan) [inline Scripts/pre-commit.sh]: Ok, but that's for you and no else no?  You leave here, but it's not that normal to add these utilities that you like to a shared repo

**Pull request merged**